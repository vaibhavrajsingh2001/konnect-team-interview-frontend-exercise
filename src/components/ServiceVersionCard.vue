<template>
  <li class="version-card">
    <div class="version-details">
      <span class="version-name">v{{ name }}</span>
      <span
        class="version-description"
        :title="description"
      >
        {{ description }}
      </span>
      <span
        class="service-type"
        :class="type === 'HTTP' ? 'http-bg-color' : 'rest-bg-color'"
      >{{ type }}</span>
    </div>
    <span
      v-if="developer"
      class="developer-details"
    >
      <k-avatar
        :avatar-url="developer.avatar"
        :email="developer.email"
        :name="developer.name"
      />
      <div>
        <p class="developer-name">{{ developer.name }}</p>
        <p class="service-updated-at">{{ fromNow(updatedAt) }}</p>
      </div>
    </span>
  </li>
</template>
<script lang="ts">
import { defineComponent, type PropType } from 'vue'

import KAvatar from '@/components/KAvatar.vue'

import { fromNow } from '@/util'
import type { Developer, Service } from '@/types'

export default defineComponent({
  name: 'ServiceVersionCard',
  components: {
    KAvatar,
  },
  props: {
    name: {
      type: String,
      required: true,
    },
    description: {
      type: String,
      required: true,
    },
    developer: {
      type: Object as PropType<Developer>,
      default: null,
    },
    updatedAt: {
      type: String,
      required: true,
    },
    type: {
      type: String as PropType<Service['type']>,
      required: true,
      validator: (value: string) => ['REST', 'HTTP'].includes(value),
    },
  },
  methods: { fromNow },
})
</script>
<style lang="scss">
.version-card {
  align-items: center;
  display: flex;
  gap: 20rem;
  justify-content: space-between;
  padding: 0.8rem 0;
}

.version-details {
  align-items: center;
  display: flex;
  gap: 3.2rem;

  .version-name {
    color: #262626;
    font-size: 1.3rem;
    font-weight: 600;
    line-height: 2.4rem;
    width: 4rem;
  }

  .version-description {
    -webkit-box-orient: vertical;
    color: #8A8A8A;
    display: -webkit-box;
    font-size: 1.2rem;
    -webkit-line-clamp: 2;
    line-height: 1.6rem;
    overflow: hidden;
    width: 25rem;
  }
}

.service-type {
  border-radius: 0.4rem;
  color: #1155CB;
  font-size: 1rem;
  font-weight: 500;
  padding: 0.6rem 0.5rem;

  &.rest-bg-color {
    background-color: #BDD3F9;
  }

  &.http-bg-color {
    background-color: #F8F8FA;
  }
}

.developer-details {
  align-items: start;
  display: flex;
  gap: 0.6rem;
  justify-content: start;
  text-align: left;

  .k-avatar {
    height: 2rem;
    width: 2rem;
  }

  .developer-name {
    color: #3C4557;
    font-size: 1.3rem;
    font-weight: 600;
    line-height: 2.4rem;
  }

  .service-updated-at {
    color: #8A8A8A;
    font-size: 1.2rem;
    font-weight: 400;
    line-height: 1.6rem;
  }
}
</style>
