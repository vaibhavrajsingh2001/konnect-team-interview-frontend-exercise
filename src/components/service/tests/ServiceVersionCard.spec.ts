import { shallowMount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import ServiceVersionCard from '@/components/service/ServiceVersionCard.vue'
import type { Developer } from '@/types'

describe('ServiceVersionCard', () => {
  it('renders correctly with no developers', () => {
    const wrapper = shallowMount(ServiceVersionCard, {
      props: {
        name: '1.0.0',
        description: 'Test description',
        type: 'HTTP',
        updatedAt: '2022-01-01T00:00:00Z',
      },
    })

    expect(wrapper.find('.version-name').text()).to.equal('v1.0.0')
    expect(wrapper.find('.version-description').text()).to.equal('Test description')
    expect(wrapper.find('.service-type').text()).to.equal('HTTP')
    expect(wrapper.find('.service-type').classes()).toContain('http-bg-color')

    // Test if developer-details is not rendered
    expect(wrapper.find('.developer-details').exists()).toBeFalsy()

    // Test snapshot
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders correctly with developers', () => {
    const developer: Developer = { id: 'e06778e0-d298-4bc7-a2fe-db9c4b371f39', name: 'Oscar Brown', email: 'Oscar.Brown87@yahoo.com', avatar: 'https://avatars.githubusercontent.com/u/55118358' }

    const wrapper = shallowMount(ServiceVersionCard, {
      props: {
        name: '1.0.0',
        description: 'Test description',
        type: 'REST',
        updatedAt: '2022-01-01T00:00:00Z',
        developer,
      },
      global: {
        stubs: {
          KAvatar: true,
        },
      },
    })

    // Test correct class is apllied when type is REST
    expect(wrapper.find('.service-type').text()).to.equal('REST')
    expect(wrapper.find('.service-type').classes()).toContain('rest-bg-color')
    // Test if developer-details is rendered
    expect(wrapper.find('.developer-details').exists()).toBeTruthy()
    expect(wrapper.find('.developer-name').text()).to.equal(developer.name)

    // Test snapshot
    expect(wrapper.html()).toMatchSnapshot()
  })
})
