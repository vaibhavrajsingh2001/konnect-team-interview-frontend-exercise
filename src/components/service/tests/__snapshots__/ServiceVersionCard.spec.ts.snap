// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ServiceVersionCard > renders correctly with developers 1`] = `
"<li class="version-card">
  <div class="version-details"><span class="version-name">v1.0.0</span><span class="version-description" title="Test description">Test description</span><span class="service-type rest-bg-color">REST</span></div><span class="developer-details"><k-avatar-stub avatarurl="https://avatars.githubusercontent.com/u/55118358" email="Oscar.Brown87@yahoo.com" name="Oscar Brown"></k-avatar-stub><div><p class="developer-name">Oscar Brown</p><p class="service-updated-at">2 years ago</p></div></span>
</li>"
`;

exports[`ServiceVersionCard > renders correctly with no developers 1`] = `
"<li class="version-card">
  <div class="version-details"><span class="version-name">v1.0.0</span><span class="version-description" title="Test description">Test description</span><span class="service-type http-bg-color">HTTP</span></div>
  <!--v-if-->
</li>"
`;
